### ISSUE TITLE
Asynchronous triggering of actions makes it possible to ack a message prior to performing the action

### ISSUE THREAD

**Issue by juliev0** [argoproj/argo-events#1977]
We try to be "exactly once" in most cases and "at least once" in the case of SIGFAULT. 

To be "exactly once" in the case of normal shutdown (SIGTERM), the logic prevents shutting down in the middle of processing a message. But the fact that actions are triggered [asynchronously](https://github.com/argoproj/argo-events/blob/master/sensors/listener.go#L335) means that we don't wait until the triggers have occurred to shut down.

Also, to be "at least once" in the case of the Sensor getting a SIGFAULT, we specifically ack the EventBus (NATS streaming as well as Jetstream) at the end of processing the message. But because the action gets triggered asynchronously, the ack can come before the action, so if the Sensor got a SIGFAULT after performing the ack and prior to triggering the action then the message would be lost.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by juliev0** [comment]
keeping alive in the hopes I'll work on it at some point

**Comment by juliev0** [comment]
@whynowy Could we add this to our project? I talked to Qianbo and he says this would be a good thing to tackle when we get a chance.

**Comment by dfarr** [comment]
@juliev0 we can take a look into supporting this

**Comment by bilalba** [comment]
In order to make it at least once, we looked into implementing a callback that would ack at the end of asynchronous trigger call completing. 

First we looked into just implementing [the ack ](https://github.com/argoproj/argo-events/blob/master/eventbus/jetstream/sensor/trigger_conn.go#L260-L268) as a callback to an asynchronous trigger completing. We would pass this callback function as an additional argument [here](https://github.com/argoproj/argo-events/blob/master/eventbus/jetstream/sensor/trigger_conn.go#L330) and forward the call back to [triggerActions()](https://github.com/argoproj/argo-events/blob/master/sensors/listener.go#L325), and [triggerwithRateLimit()](https://github.com/argoproj/argo-events/blob/master/sensors/listener.go#L325), and finally call this callback in triggerWithRateLimit. 

However, two things may happen from this. 

1. [clearAllDependencies()](https://github.com/argoproj/argo-events/blob/master/eventbus/jetstream/sensor/trigger_conn.go#L450) will get invoked even though the message has not been acked. If clearAllDependencies gets invoked, and the system shuts down, we lose out on previous messages.

2. The message will continue to be sent again if the waiting period has passed.

We then also explored having clearAllDependencies as a callback too after the trigger invocation has completed. This wouldn't work either because we will continue to listen for more events and then call clearAllDependencies at the end of a trigger, so we'll lose out on the messages that came in between.

**Comment by juliev0** [comment]
These are good findings.

It does seem like the correct (safe) order for "at least once" would be something like this, right?:
1. perform action
2. clear store
3. ack

**Comment by dfarr** [comment]
> These are good findings.
> 
> It does seem like the correct (safe) order for "at least once" would be something like this, right?:
> 
> 1. perform action
> 2. clear store
> 3. ack

Yes. In addition to this, we could optionally support at-most-once processing. I imagine this could be a configuration on either the overall sensor, or on individual triggers. The difference is in the ordering of our steps.

1. clear store
2. ack
3. perform action (could be async)

**Comment by juliev0** [comment]
I think maybe we can revisit whether we really need asynchronicity in the trigger action.
